BUG: properly close files opened by parsers

agraboso opened the pull request:
- [x] closes #13932 
- [x] tests passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

Addresses all ~~most~~ of the `ResourceWarning`s from #13932. There were three tests that opened files but did not close them. The rest were due to parsers (`CParserWrapper` and `PythonParser`) opening files or streams but never closing them. There was one case &mdash; the `mmap` one &mdash; that was tricky because there was a file handle going into it and immediately reaching a refcount of zero.

I am at a loss about writing new tests for this. Is it even possible?

agraboso committed:
BUG: properly close files opened by parsers

agraboso commented:
Argh, there are still some `ResourceWarning`s left... Let me investigate.

jreback commented:
@agraboso looks a lot cleaner! https://travis-ci.org/pydata/pandas/jobs/150768150

you can leave the mmap one (if its not easy to fix). If you can get the stata ones would be great!.

jreback commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -972,3 +972,5 @@ Bug Fixes
 - Bug in ``Index`` raises ``KeyError`` displaying incorrect column when column is not in the df and columns contains duplicate values (:issue:`13822`)
 - Bug in ``Period`` and ``PeriodIndex`` creating wrong dates when frequency has combined offset aliases (:issue:`13874`)
+
+- Bug where files opened by parsers (used by `pd.read_csv`, e.g.) were not closed properly. (:issue:`13940`)
```
double back ticks around ``pd.read_csv``. Put this next to the other csv bug fixes.

jreback replied:
say *when* these were not closed properlty (e.g. nrows not None, when else?)

jreback replied:
are we *sure* that on iteration (e.g. chunksize / iterator specified) we are closing properly? (e.g. I am not sure if it closes, then re-opens each iteration).

cc @gfyoung do you know?

gfyoung replied:
It does not re-open after each iteration AFAIK.  You can see for yourself by writing to a CSV file, opening it with `read_csv` and passing in `chunksize`.  Try then deleting the file in a separate window.  You should be able to do so.

agraboso replied:
Indeed, when getting an iterator (`chunksize` or `iterator` specified) we [do not close](https://github.com/agraboso/pandas/blob/509e2e628d8447198ef90bd63c8912e434beafa4/pandas/io/parsers.py#L399-L400) the file. When the iterator has been consumed and raises `StopIteration`, we [close](https://github.com/agraboso/pandas/blob/509e2e628d8447198ef90bd63c8912e434beafa4/pandas/io/parsers.py#L904-L906) the file. If a user requests an iterator but does not consume it fully, s/he must call [.close()](https://github.com/agraboso/pandas/blob/509e2e628d8447198ef90bd63c8912e434beafa4/pandas/io/parsers.py#L733-L734) on it &mdash; though we could define a `__del__` and take care of that.

gfyoung commented on pandas/io/parsers.py:
```diff
@@ -727,10 +731,7 @@ def __init__(self, f, engine=None, **kwds):
-            self._engine._reader.close()
-        except:
-            pass
+        self._engine.close()
```
Why can we get rid of this `try`-`except` block?

agraboso replied:
I shouldn't have. But this was only used by `CParserWrapper`, so I have put it there now in my last commit.

agraboso committed:
Properly close opened files in three tests

agraboso committed:
Properly close opened files in StataReader

agraboso committed:
Properly close opened files in StataWriter

agraboso committed:
Fix whatsnew entries

agraboso committed:
On close, CParserWrapper must call self._reader.close()

agraboso commented:
@jreback The `mmap` one is fixed since the first commit &mdash; I was just making the point that it took me a while to understand what was wrong.

I have fixed the `stata` cases too:
- `StataWriter` was easy: I moved the opening of the file to the [`write_file`](https://github.com/agraboso/pandas/blob/509e2e628d8447198ef90bd63c8912e434beafa4/pandas/io/stata.py#L2078-L2093) method. The file is thus always opened and closed within it, whether errors are raised or not.
- `StataReader` did not close the file when raising an error. I fixed it by wrapping most of the logic in its [`read`](https://github.com/agraboso/pandas/blob/509e2e628d8447198ef90bd63c8912e434beafa4/pandas/io/stata.py#L1406) method in a `try/except` block. This even takes care of closing the file when a `StopIteration` is raised.

agraboso committed:
Fix long line

jreback commented on pandas/io/stata.py:
```diff
@@ -1411,150 +1407,155 @@ def read(self, nrows=None, convert_dates=None,
-        if read_len <= 0:
-            # Iterator has finished, should never be here unless
-            # we are reading the file incrementally
+        try:
```
you don't nee dthe try/except over everthing here. pls narrow down as much as possible.

agraboso replied:
I narrowed it down to the few places where an explicit `ValueError` might come from another `StataReader` method. It catches those cases that are in the tests, but it is possible for other lines to raise an error that is not caught &mdash; which would leave an open file handler.

jreback commented on pandas/io/stata.py:
```diff
@@ -2078,16 +2077,21 @@ def _prepare_pandas(self, data):
+        )
+        try:
+            self._write_header(time_stamp=self._time_stamp,
+                               data_label=self._data_label)
```
why do you need this block?

agraboso replied:
Both `self._write_header` and `self._write_variable_labels()` may themselves raise a `ValueError`. The other method calls may also raise errors bubbling up from calls to other functions. In any and all cases, we want to avoid leaving `self._file` open.

agraboso committed:
Make try/except blocks in StataReader.read as small as possible

jreback commented on pandas/io/stata.py:
```diff
@@ -1494,7 +1491,11 @@ def read(self, nrows=None, convert_dates=None,
 
         if columns is not None:
-            data = self._do_select_columns(data, columns)
+            try:
```
it might be worth it to create a context manager so this is a bit easier, eg.

```
with close_on_error():
    data = self._do_select_columns(data, columns)
```

jreback replied:
actually you only have 2 uses of this (in the main body), so may not be worht it.

agraboso replied:
Yeah, that was my thought as soon as I saw your previous comment. I'm open to implementing it if you want it but, as you say, it's probably not worth it.

agraboso replied:
Another possibility &mdash; though, given your aversion to my wrapping most of `StataReader.read` in a `try/except` block, you might not like it &mdash; is writing a `close_on_error` decorator and wrapping `StataReader.read` &mdash; and maybe `StataWriter.write_file` too &mdash; in it.

jreback commented:
I don't want to just wrap things at a high level, these should be as granular as possible.These can be actual errors (or errors in the tests, its hard to tell). Let uncaught resource warnings show thru (and then can investigate these individually).

agraboso commented:
Notice that my proposal of wrapping most of `StataReader.read` in a `try/except` block had both a `self.close()` and a `raise`, so that we would still see all warnings and errors while making sure we do close the file.

agraboso committed:
Fix rebase

agraboso committed:
Fix linting error

agraboso committed:
Properly close opened files in two tests

agraboso committed:
Properly close opened files in XportReader and SAS7BDATReader

agraboso committed:
Do not acquire list as file handler to close

agraboso committed:
Close open files in TextFileReader upon StopIteration

jreback commented on pandas/io/sas/sas_xport.py:
```diff
@@ -253,6 +253,9 @@ def __init__(self, filepath_or_buffer, index=None, encoding='ISO-8859-1',
         self._read_header()
 
+    def close(self):
+        self.filepath_or_buffer.close()
```
is this always a filehandle-like?

agraboso replied:
It's either a file handle coming from [L244](https://github.com/agraboso/pandas/blob/3fa7d25a00372a1312cd8fad0a82082e997f51a4/pandas/io/sas/sas_xport.py#L244) or a `BytesIO` object coming from [L252](https://github.com/agraboso/pandas/blob/3fa7d25a00372a1312cd8fad0a82082e997f51a4/pandas/io/sas/sas_xport.py#L252). So it is always safe to close it.

jreback commented:
hah! this turned into a rabbit hole!. lmk when you are ready.

agraboso commented:
> hah! this turned into a rabbit hole!

Tell me about it! It's possible that other parsers leave open files too, but I'd leave to for another PR. I think this one is ready.

 lmk when you are ready.

And it's all green!

jreback commented:
thanks @agraboso I merged. left a couple comments that might be relevant for next PR :)

we need to try to fix remaining ``ResourceWarnings`` then raise if ANY are detected (its actually pretty easy). in ``pandas.util.testing.TestCase.setUpClass`` you can set the warning to error. This way its not set globally when ``testing_mode()`` is called (in case testing is just imported by a user and not actually run in the test suite).

jreback commented on pandas/io/parsers.py:
```diff
@@ -393,11 +393,15 @@ def _read(filepath_or_buffer, kwds):
                                   " together yet.")
     elif nrows is not None:
-        return parser.read(nrows)
+        data = parser.read(nrows)
```
if you do

```
try:
    return parser.read(nrows)
finally:
    parse.close()
```

I think you can eliminate the ``.close()`` calls where you raise on an exception (also needs changing at L402).
and slightly more idiomatic.

jreback commented on pandas/io/parsers.py:
```diff
@@ -393,11 +393,15 @@ def _read(filepath_or_buffer, kwds):
 
-    return parser.read()
+    data = parser.read()
+    parser.close()
```
e.g here
